include/dos_inc.h

```diff
@@ -366,25 +366,59 @@ constexpr PhysPt assert_macro_args_ok()
 	return 0;
 }
 
-#define VERIFY_SSET_ARGS(n, s, f, v)                                           \
+#define VERIFY_SSET_ARGS(n, s, f, v) \
 	assert_macro_args_ok<n, s, decltype(s::f), decltype(v)>()
-#define VERIFY_SGET_ARGS(n, s, f)                                              \
+#define VERIFY_SGET_ARGS(n, s, f) \
 	assert_macro_args_ok<n, s, decltype(s::f)>()
 
-#define SSET_BYTE(s, f, v)                                                     \
+#define SSET_BYTE(s, f, v) \
 	mem_writeb(VERIFY_SSET_ARGS(1, s, f, v) + pt + offsetof(s, f), v)
-#define SSET_WORD(s, f, v)                                                     \
+#define SSET_WORD(s, f, v) \
 	mem_writew(VERIFY_SSET_ARGS(2, s, f, v) + pt + offsetof(s, f), v)
-#define SSET_DWORD(s, f, v)                                                    \
+#define SSET_DWORD(s, f, v) \
 	mem_writed(VERIFY_SSET_ARGS(4, s, f, v) + pt + offsetof(s, f), v)
 
-#define SGET_BYTE(s, f)                                                        \
+#define SGET_BYTE(s, f) \
 	mem_readb(VERIFY_SGET_ARGS(1, s, f) + pt + offsetof(s, f))
-#define SGET_WORD(s, f)                                                        \
+#define SGET_WORD(s, f) \
 	mem_readw(VERIFY_SGET_ARGS(2, s, f) + pt + offsetof(s, f))
-#define SGET_DWORD(s, f)                                                       \
+#define SGET_DWORD(s, f) \
 	mem_readd(VERIFY_SGET_ARGS(4, s, f) + pt + offsetof(s, f))
 
+/* Macros SSET_*_ARRAY and SGET_*_ARRAY are similar to SSET_* and SGET_*, but
+ * they are to be used for arrays. Index range is not being checked.
+ *
+ * Example usage:
+ *
+ *   SSET_WORD_ARRAY(dos-structure-name, field-name, array-index, value);
+ *   uint16_t x = SGET_WORD_ARRAY(dos-structure-name, field-name, array-index);
+ */
+
+#define VERIFY_SSET_ARRAY_ARGS(n, s, f, v) \
+	assert_macro_args_ok<n, s, decltype(*s::f), decltype(v)>()
+#define VERIFY_SGET_ARRAY_ARGS(n, s, f) \
+	assert_macro_args_ok<n, s, decltype(*s::f)>()
+
+#define SSET_BYTE_ARRAY(s, f, i, v) \
+	mem_writeb(VERIFY_SSET_ARRAY_ARGS(1, s, f, v) + \
+	           static_cast<uint32_t>(pt + offsetof(s, f) + 1 * i), v)
+#define SSET_WORD_ARRAY(s, f, i, v) \
+	mem_writew(VERIFY_SSET_ARRAY_ARGS(2, s, f, v) + \
+	           static_cast<uint32_t>(pt + offsetof(s, f) + 2 * i), v)
+#define SSET_DWORD_ARRAY(s, f, i, v) \
+	mem_writed(VERIFY_SSET_ARRAY_ARGS(4, s, f, v) + \
+	           static_cast<uint32_t>(pt + offsetof(s, f) + 4 * i), v)
+
+#define SGET_BYTE_ARRAY(s, f, i) \
+	mem_readb(VERIFY_SGET_ARRAY_ARGS(1, s, f) + \
+	          static_cast<uint32_t>(pt + offsetof(s, f) + 1 * i))
+#define SGET_WORD_ARRAY(s, f, i) \
+	mem_readw(VERIFY_SGET_ARRAY_ARGS(2, s, f) + \
+	          static_cast<uint32_t>(pt + offsetof(s, f) + 2 * i))
+#define SGET_DWORD_ARRAY(s, f, i) \
+	mem_readd(VERIFY_SGET_ARRAY_ARGS(4, s, f) + \
+	          static_cast<uint32_t>(pt + offsetof(s, f) + 4 * i))
+
```

Comment:
Huh, do we need tests for these dirty macos? 😆

Reply:
Do you mean unit tests?

Reply:
Yes.

Reply:
Thanks, good stuff!